File: 2022/python_solutions/4.py
from dataclasses import dataclass

data = """2-4,6-8
2-3,4-5
5-7,7-9
2-8,3-7
6-6,4-6
2-6,4-8"""

data = open("../inputs/4.txt").read()


@dataclass
class Assignment:
    start: int
    stop: int

    def contains(self, other: 'Assignment'):
        return self.start <= other.start and other.stop <= self.stop

    def overlaps(self, other: 'Assignment'):
        # my solution
        # return self.contains(other) or any(x for x in range(self.start, self.stop+1) if x in range(other.start, other.stop + 1))

        # Solution from https://youtu.be/15qPSEFoR0U?t=202
        return not (self.stop < other.start or other.stop < self.start)

    @staticmethod
    def parse(s: str):
        return Assignment(*map(int, s.split('-')))


assignments = [[Assignment.parse(x) for x in pair.split(",")]
               for pair in data.splitlines()]


def check(func): return lambda pair: func(
    pair[0], pair[1]) or func(pair[1], pair[0])


print(len(list(filter(check(Assignment.contains), assignments))))
print(len(list(filter(check(Assignment.overlaps), assignments))))

File: 2022/haskell_solutions/4.hs
{-# LANGUAGE OverloadedStrings #-}

import Data.List (sort)
import Data.Text qualified as T

-- data Assignment = Assignment {start :: Int, stop :: Int} deriving (Show)
data Assignment = Assignment Int Int deriving (Show)

start (Assignment start _) = start

stop (Assignment _ stop) = stop

type Pair = (Assignment, Assignment)

parseInt :: String -> Int
parseInt = read

contains x y = (start x <= start y) && (stop y <= stop x)

overlaps x y = not $ (stop x < start y) || (stop y < start x)

parseAssignment x = Assignment start stop
  where
    [start, stop] = map (read . T.unpack) $ T.splitOn "-" x

main = do
  --   pairs <- T.lines . T.pack <$> readFile "../inputs/4.sample"
  pairs <- T.lines . T.pack <$> readFile "../inputs/4.txt"
  print $ solve contains pairs
  print $ solve overlaps pairs
  where
    solve f = length . filter (\[x, y] -> f x y || f y x) . map (map parseAssignment . T.splitOn ",")